Add protocol buffers for binary injection/joining

bg451 opened the pull request:
* Replace the internal use of `encoding/binary` inside of `(*splitBinaryPropagator).Join` and `(*splitBinaryPropagator).Inject` with Google's Protocol buffers. Gogoproto is used for the go code generation since it provides considerable performance gains over vanilla protobufs.

* Adds a ProtobufCarrier to allow end users to use other protobuf encoding formats, i.e. jsonpb.

The benchmarks were run on an AWS m3.large. 
```
name                                           old time/op    new time/op    delta
Span_Empty                                        710ns ±55%     657ns ± 0%   -7.53%          (p=0.037 n=8+7)
Span_100Events                                   48.6µs ± 1%    49.9µs ± 0%   +2.66%         (p=0.000 n=9+10)
Span_1000Events                                  48.5µs ± 0%    49.9µs ± 0%   +2.85%        (p=0.000 n=10+10)
Span_100Tags                                     47.0µs ± 1%    46.5µs ± 2%   -0.91%        (p=0.023 n=10+10)
Span_1000Tags                                    46.8µs ± 1%    47.0µs ± 1%     ~           (p=0.529 n=10+10)
Span_100BaggageItems                              155µs ± 0%     157µs ± 0%   +1.30%        (p=0.000 n=10+10)
TrimmedSpan_100Events_100Tags_100BaggageItems     181µs ± 0%     181µs ± 0%   +0.36%        (p=0.000 n=10+10)
Inject_SplitText_Empty                            468ns ± 1%     476ns ± 1%   +1.56%        (p=0.000 n=10+10)
Inject_SplitText_100BaggageItems                 20.0µs ± 2%    20.1µs ± 1%     ~           (p=0.089 n=10+10)
Inject_GoHTTPHeader_Empty                        2.98µs ± 4%    1.39µs ± 1%  -53.33%        (p=0.000 n=10+10)
Inject_GoHTTPHeader_100BaggageItems              70.7µs ± 1%    48.4µs ± 1%  -31.57%          (p=0.000 n=8+9)
Inject_SplitBinary_Empty                         1.30µs ± 2%    0.31µs ± 1%  -76.08%         (p=0.000 n=10+9)
Inject_SplitBinary_100BaggageItems               52.8µs ± 1%    36.0µs ± 1%  -31.94%        (p=0.000 n=10+10)
Join_SplitText_Empty                             1.61µs ± 0%    1.62µs ± 1%   +0.86%        (p=0.000 n=10+10)
Join_SplitText_100BaggageItems                   54.0µs ± 2%    53.8µs ± 1%     ~           (p=0.739 n=10+10)
Join_GoHTTPHeader_Empty                          2.52µs ± 1%    1.62µs ± 0%  -35.97%         (p=0.000 n=10+8)
Join_GoHTTPHeader_100BaggageItems                 152µs ± 2%      91µs ± 2%  -40.10%        (p=0.000 n=10+10)
Join_SplitBinary_Empty                           1.73µs ± 1%    0.96µs ± 1%  -44.35%        (p=0.000 n=10+10)
Join_SplitBinary_100BaggageItems                 96.8µs ± 1%    42.4µs ± 4%  -56.15%         (p=0.000 n=9+10)

name                                           old alloc/op   new alloc/op   delta
Span_Empty                                       0.75B ±700%    0.00B ±NaN%     ~             (p=1.059 n=8+9)
Span_100Events                                   16.1kB ± 0%    16.1kB ± 0%     ~     (all samples are equal)
Span_1000Events                                  16.1kB ± 0%    16.1kB ± 0%     ~     (all samples are equal)
Span_100Tags                                     10.5kB ± 0%    10.5kB ± 0%     ~           (p=0.355 n=10+10)
Span_1000Tags                                    10.5kB ± 0%    10.5kB ± 0%     ~           (p=0.692 n=10+10)
Span_100BaggageItems                             10.5kB ± 0%    10.5kB ± 0%     ~           (p=0.724 n=10+10)
TrimmedSpan_100Events_100Tags_100BaggageItems    0.00B ±NaN%    0.00B ±NaN%     ~     (all samples are equal)
Inject_SplitText_Empty                            48.0B ± 0%     48.0B ± 0%     ~     (all samples are equal)
Inject_SplitText_100BaggageItems                 3.25kB ± 0%    3.25kB ± 0%     ~     (all samples are equal)
Inject_GoHTTPHeader_Empty                          641B ± 0%      321B ± 0%  -49.92%        (p=0.000 n=10+10)
Inject_GoHTTPHeader_100BaggageItems              30.1kB ± 0%    18.9kB ± 0%  -37.18%         (p=0.000 n=9+10)
Inject_SplitBinary_Empty                           368B ± 0%       80B ± 0%  -78.26%        (p=0.000 n=10+10)
Inject_SplitBinary_100BaggageItems               6.77kB ± 0%    6.39kB ± 0%   -5.56%        (p=0.000 n=10+10)
Join_SplitText_Empty                              16.0B ± 0%     16.0B ± 0%     ~     (all samples are equal)
Join_SplitText_100BaggageItems                   10.5kB ± 0%    10.5kB ± 0%     ~            (p=0.823 n=10+9)
Join_GoHTTPHeader_Empty                            288B ± 0%       80B ± 0%  -72.22%        (p=0.000 n=10+10)
Join_GoHTTPHeader_100BaggageItems                33.3kB ± 0%    27.8kB ± 0%  -16.57%        (p=0.000 n=10+10)
Join_SplitBinary_Empty                             240B ± 0%       48B ± 0%  -80.00%        (p=0.000 n=10+10)
Join_SplitBinary_100BaggageItems                 22.5kB ± 0%    17.0kB ± 0%  -24.52%        (p=0.000 n=10+10)

name                                           old allocs/op  new allocs/op  delta
Span_Empty                                        0.00 ±NaN%     0.00 ±NaN%     ~     (all samples are equal)
Span_100Events                                     8.00 ± 0%      8.00 ± 0%     ~     (all samples are equal)
Span_1000Events                                    8.00 ± 0%      8.00 ± 0%     ~     (all samples are equal)
Span_100Tags                                       11.0 ± 0%      11.0 ± 0%     ~     (all samples are equal)
Span_1000Tags                                      11.0 ± 0%      11.0 ± 0%     ~     (all samples are equal)
Span_100BaggageItems                               11.0 ± 0%      11.0 ± 0%     ~     (all samples are equal)
TrimmedSpan_100Events_100Tags_100BaggageItems     0.00 ±NaN%     0.00 ±NaN%     ~     (all samples are equal)
Inject_SplitText_Empty                             3.00 ± 0%      3.00 ± 0%     ~     (all samples are equal)
Inject_SplitText_100BaggageItems                    103 ± 0%       103 ± 0%     ~     (all samples are equal)
  ject_GoHTTPHeader_Empty                          15.0 ± 0%       5.0 ± 0%  -66.67%        (p=0.000 n=10+10)
Inject_GoHTTPHeader_100BaggageItems                 421 ± 0%        11 ± 0%  -97.39%        (p=0.000 n=10+10)
Inject_SplitBinary_Empty                           11.0 ± 0%       3.0 ± 0%  -72.73%        (p=0.000 n=10+10)
Inject_SplitBinary_100BaggageItems                  411 ± 0%         7 ± 0%  -98.30%        (p=0.000 n=10+10)
Join_SplitText_Empty                               1.00 ± 0%      1.00 ± 0%     ~     (all samples are equal)
Join_SplitText_100BaggageItems                     12.0 ± 0%      12.0 ± 0%     ~     (all samples are equal)
Join_GoHTTPHeader_Empty                            13.0 ± 0%       3.0 ± 0%  -76.92%        (p=0.000 n=10+10)
Join_GoHTTPHeader_100BaggageItems                   622 ± 0%       217 ± 0%  -65.11%        (p=0.000 n=10+10)
Join_SplitBinary_Empty                             11.0 ± 0%       2.0 ± 0%  -81.82%        (p=0.000 n=10+10)
Join_SplitBinary_100BaggageItems                    619 ± 0%       214 ± 0%  -65.43%        (p=0.000 n=10+10)
```

bensigelman commented on wire/wire.proto:
```diff
@@ -0,0 +1,13 @@
+option go_package = "wire";
+
+message TracerState {
+  int64 trace_id = 1;
```
nit: `fixed64` will be better here and on the line below. (["Always eight bytes. More efficient than uint64 if values are often greater than 256."](https://developers.google.com/protocol-buffers/docs/proto3#scalar))

bensigelman commented on wire/wire.proto:
```diff
@@ -0,0 +1,13 @@
+syntax = "proto3";
+package basictracer_go.wire;
```
Thinking out loud: since the binary (and text) protocol should be shared across *all* "basictracer" impls (not just Go), eventually this .proto file should be in some sort of `basictracer-common` repo. That can/should wait until things have stabilized more, though.

yurishkuro replied:
this is a Dapper-ish impl though, not general purpose. E.g wouldn't work for Zipkin.

bensigelman replied:
Certainly true... `basictracer` is already making assumptions about which tracer state is worth keeping around, though. (FWIW, it wouldn't quite work for dapper, either: there were a few other (literal) bits sent over the wire, not just a `sampled` bool)

tschottdorf replied:
agree that the proto should move outside of the main repo. I'd do that rather sooner than later so that the other impls can catch up.

bensigelman commented on wire/wire.proto:
```diff
@@ -0,0 +1,13 @@
+package basictracer_go.wire;
+option go_package = "wire";
+
+message TracerState {
```
Per https://github.com/opentracing/opentracing.github.io/issues/73#issuecomment-193499918, the semantics of the split between "tracer state" and "baggage" is up in the air (and – unless someone objects – I suspect they will be unified). Not sure if we should (a) wait for https://github.com/opentracing/opentracing.github.io/issues/73 to resolve before merging this or (b) go with what you have here with the understanding that it's going to change imminently.

bg451 replied:
I'm going to wait for opentracing/opentracing.github.io#73 to resolve first, since it seems that the github timer is going to start soon

bensigelman commented on wire/carrier.go:
```diff
@@ -0,0 +1,38 @@
+// the underlying datastructur. The reason for implementing DelagatingCarrier
+// is to allow for users to serialize the underlying protocol buffers using
+// jsonpb or any other serialization forms they want.
+type ProtobufCarrier struct {
```
I'm not reading this as carefully as I perhaps should, but am I right in understanding that this DelegatingCarrier impl is not actually used anywhere (yet)?

bg451 replied:
It's not, I figured I'd go ahead and include it though since we're already providing the generated protobuf code.

tschottdorf replied:
https://github.com/cockroachdb/cockroach/blob/master/util/tracing/span.go


On Mon, Mar 7, 2016 at 11:16 PM bhs <notifications@github.com> wrote:

> In wire/carrier.go
> <https://github.com/opentracing/basictracer-go/pull/13#discussion_r55314484>
> :
>
> > @@ -0,0 +1,38 @@
> > +package wire
> > +
> > +// ProtobufCarrier is a DelegatingCarrier that uses protocol buffers as the
> > +// the underlying datastructur. The reason for implementing DelagatingCarrier
> > +// is to allow for users to serialize the underlying protocol buffers using
> > +// jsonpb or any other serialization forms they want.
> > +type ProtobufCarrier struct {
>
> I'm not reading this as carefully as I perhaps should, but am I right in
> understanding that this DelegatingCarrier impl is not actually used
> anywhere (yet)?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opentracing/basictracer-go/pull/13/files#r55314484>.
>

bensigelman commented:
```
Inject_GoHTTPHeader_100BaggageItems                 421 ± 0%        11 ± 0%  -97.39%        (p=0.000 n=10+10)
```

^^^ nice!!

bg451 commented:
On a side note, does having the protobuf code in the package `wire` make sense? Something about it just feels off to me, but I'm not sure why.

bensigelman commented:
> On a side note, does having the protobuf code in the package `wire` make sense? Something about it just feels off to me, but I'm not sure why.

Per that comment I made, I think the .proto belongs somewhere far away (probably in a different repo altogether)... I can see lots of alternatives to `wire` but I was having trouble coming up with strong arguments one way or the other.

bg451 commented:
I agree that it should be moved to something like `basictracer-common`. I suppose for now we can leave it in here until OT 1.0 becomes official or someone starts working on another basictracer implementation.

tschottdorf commented on wire/carrier.go:
```diff
@@ -0,0 +1,38 @@
+package wire
+
+// ProtobufCarrier is a DelegatingCarrier that uses protocol buffers as the
```
can't find the comment now, but DelegatingCarrier is already being used: https://github.com/cockroachdb/cockroach/blob/master/util/tracing/span.go

bg451 commented:
Rebased and squashed everything now that #15 is in.

Here's the updated performance numbers:
```
name                                             old time/op    new time/op    delta
Inject_Binary_Empty-2                              1.03µs ± 0%    0.49µs ± 0%  -51.95%         (p=0.000 n=10+7)
Inject_Binary_100BaggageItems-2                    61.0µs ± 1%    16.0µs ± 2%  -73.85%         (p=0.000 n=10+9)
Join_Binary_Empty-2                                1.75µs ± 1%    1.49µs ± 0%  -15.14%        (p=0.000 n=10+10)
Join_Binary_100BaggageItems-2                      95.4µs ± 0%    46.6µs ± 1%  -51.18%        (p=0.000 n=10+10)

name                                             old alloc/op   new alloc/op   delta
Inject_Binary_Empty-2                                219B ± 0%      168B ± 0%  -23.29%        (p=0.000 n=10+10)
Inject_Binary_100BaggageItems-2                    23.8kB ± 1%    16.3kB ± 0%  -31.34%        (p=0.000 n=10+10)
Join_Binary_Empty-2                                  256B ± 0%      224B ± 0%  -12.50%        (p=0.000 n=10+10)
Join_Binary_100BaggageItems-2                      22.5kB ± 0%    21.7kB ± 0%   -3.49%         (p=0.000 n=10+9)

name                                             old allocs/op  new allocs/op  delta
Inject_Binary_Empty-2                                9.00 ± 0%      5.00 ± 0%  -44.44%        (p=0.000 n=10+10)
Inject_Binary_100BaggageItems-2                       409 ± 0%         5 ± 0%  -98.78%        (p=0.000 n=10+10)
Join_Binary_Empty-2                                  10.0 ± 0%       6.0 ± 0%  -40.00%        (p=0.000 n=10+10)
Join_Binary_100BaggageItems-2                         618 ± 0%       217 ± 0%  -64.89%        (p=0.000 n=10+10)

bg451 commented:
cc @bensigelman @tschottdorf 

Is there any way to get `golint` to ignore specific files? It's not happy with the generated protobuf code.

bensigelman commented on propagation_ot.go:
```diff
@@ -137,50 +139,26 @@ func (p *binaryPropagator) Inject(
-	if err != nil {
+	// Write the length of the marshalled binary to the writer.
+	len := uint32(len(b))
+	if err := binary.Write(carrier, binary.BigEndian, &len); err != nil {
```
mostly just curious: why is the `&` necessary here?

bg451 replied:
It's not. Interestingly, passing a pointer is marginally faster than by value.

```
name                     old time/op    new time/op    delta
Inject_Binary_Empty-4     456ns ± 1%     392ns ± 3%  -14.01%         (p=0.000 n=9+10)
```

tschottdorf replied:
Use `binary.PutVarint` instead of `Write` and you can avoid throwing `&len` on the stack (where it likely lives now).

bensigelman commented:
@bg451 basically LGTM with a few caveats:
- we will need to move the proto to `basictracer-common` when we have a second `basictracer` platform
- there's the silly mechanical issue of getting TravisCI to ignore the genfiles
- I am wondering about how to represent the "sampled" state in a forwards-compatible manner...

bensigelman commented on wire/wire.proto:
```diff
@@ -0,0 +1,10 @@
+message TracerState {
+  sfixed64 trace_id = 1;
+  sfixed64 span_id = 2;
+  bool    sampled = 3;
```
Part of me wants to make this an int32 and use the low bit to represent the sampled bool proper... In order to draw statistical conclusions from sampled trace data it is helpful to record and propagate the effective sampling rate. It's also helpful to record if the trace sampling was "forced" (a la `SAMPLING_PRIORITY` here: http://opentracing.io/use-cases/#setting-sampling-priority-before-the-trace-starts) as a bit here. Oh well.

Having written down the above, I'm fine with leaving this as a bool for now, too.

bg451 replied:
I'll go ahead and open an issue to have this documented for later discussion.

bensigelman commented:
This LGTM, though it would be nice if @tschottdorf took a look since he is using `basictracer-go` at cockroach.

bensigelman commented on Makefile:
```diff
@@ -11,9 +11,10 @@ bench:
+	# Ignore grep's exit code since no match returns 1.
+	-golint ./... | grep --invert-match -E '^.*\.pb\.go'
+	@
+	@! (golint ./... |grep --invert-match -E '^.*\.pb\.go' | read dummy)
```
Sheesh! :)

I think there's a way to do this with `find`, but life is too short...

tschottdorf replied:
From our `Makefile` (similarly nasty):

```
@! golint $(PKG) | \
	  grep -vE '(\.pb\.go|\.pb\.gw\.go|embedded\.go|_string\.go|LastInsertId|sql/parser/(yaccpar|sql\.y):)'
```

tschottdorf commented on propagation_ot.go:
```diff
@@ -137,50 +139,26 @@ func (p *binaryPropagator) Inject(
-	err = binary.Write(carrier, binary.BigEndian, int32(len(sc.raw.Baggage)))
-	if err != nil {
+	// Write the length of the marshalled binary to the writer.
+	len := uint32(len(b))
```
Rename this variable to something that doesn't shadow a built-in.

tschottdorf commented on propagation_ot.go:
```diff
@@ -191,66 +169,39 @@ func (p *binaryPropagator) Join(
+	// to read. By reading the length, we can allocate a fixed sized buf and read
+	// the exact amount of bytes into it.
+	var len uint32
+	if err := binary.Read(carrier, binary.BigEndian, &len); err != nil {
```
`binary.ReadUVarint`

bensigelman replied:
Only caveat here is that we want these encodings to work across languages and many platforms do not have nice `binary.ReadUVarint` encoding/decoding intrinsically.

Meta-caveat would be that we are no binding ourselves to platforms that at least have protobuf support, at which point they must *somehow* have access to consistent varint encoding here.

... But this brings up a good point. Since we need this to be portable across platforms and we are already binding to protobufs, better to use something like https://github.com/gogo/protobuf/blob/200875106f3bf0eb01eb297dae30b250a25ffc84/proto/decode.go#L63 directly?

bensigelman replied:
(... or just stick with bigendian encoding, which at least is totally portable)

tschottdorf replied:
SGTM, sticking to the proto encoding seems sane.

bg451 replied:
Using the proto encoding won't work since it returns a variable size slice, so you don't know how many bytes to initially read on the Join side. I think sticking with the current fixed-width, big endian version is the way to go.

tschottdorf replied:
I thought it had a fixed-size encoding to. At least there are `sfixed64` (or something like that) data types.

bg451 replied:
I might be misreading the comments, but writing the size of the marshalled Protobuf message is mainly a performance optimization to prevent allocating extra space. If you encode that size using the builtin proto function `EncodeVarint` that Ben linked, the returned `[]byte` has a varying size. Check out this line https://github.com/gogo/protobuf/blob/200875106f3bf0eb01eb297dae30b250a25ffc84/proto/encode.go#L92.

bensigelman replied:
@bg451 this is complex enough that I think I'm sold on `fixed32`... That said, I have a few thoughts about this...

Really we should be talking about the decode – not encode – function here:

https://github.com/gogo/protobuf/blob/200875106f3bf0eb01eb297dae30b250a25ffc84/proto/decode.go#L57

```
// DecodeVarint reads a varint-encoded integer from the slice.
// It returns the integer and the number of bytes consumed, or
// zero if there is not enough.
// This is the format for the
// int32, int64, uint32, uint64, bool, and enum
// protocol buffer types.
func DecodeVarint(buf []byte) (x uint64, n int) { ... }
```

So, per the "if there's not enough" bit, you could read progressively longer byte slices until you're able to decode the size. Or there's a more streaming-ish impl. It basically looks a lot like reading UTF-8 chars, etc... easy to do in a streaming fashion.

But yeah: I'm fine with punting and doing fixed32 for the sake of simplicity.

bg451 replied:
It's definitely something worth revisiting down the line when other basictracer implementations come around.

tschottdorf commented on propagation_ot.go:
```diff
@@ -191,66 +169,39 @@ func (p *binaryPropagator) Join(
+	// since it keeps resizing the underlying buffer as it encounters more bytes
+	// to read. By reading the length, we can allocate a fixed sized buf and read
+	// the exact amount of bytes into it.
+	var len uint32
```
Ditto about not shadowing `len`.

tschottdorf commented on propagation_ot.go:
```diff
@@ -191,66 +169,39 @@ func (p *binaryPropagator) Join(
 	}
-	if err := binary.Read(carrier, binary.BigEndian, &sampledByte); err != nil {
+	buf := make([]byte, len)
+	if n, err := carrier.Read(buf); err != nil {
```
What's the underlying carrier really? If it's already a byte-backed type, it'd be better here to not have to make a new slice and copy everything over just to unmarshal.

tschottdorf commented:
LGTM, but needs a rebase and squash.

bg451 commented:
PTAL. Comments addressed, squashed and rebased.

bensigelman commented:
LGTM though I think we should give @tschottdorf a chance to take a look since he had mentioned the varint thing.

tschottdorf commented:
All good, nothing I'm concerned about.

On Thu, Mar 24, 2016, 19:52 bhs <notifications@github.com> wrote:

> LGTM though I think we should give @tschottdorf
> <https://github.com/tschottdorf> a chance to take a look since he had
> mentioned the varint thing.
>
> —
> You are receiving this because you were mentioned.
>
> Reply to this email directly or view it on GitHub
> <https://github.com/opentracing/basictracer-go/pull/13#issuecomment-201076582>
>
